### ISSUE TITLE
Pre-admission estimation of reader memory consumtion is static

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#4297]
We restrict the concurrency of reads, to avoid consuming an unbounded amount of memory. Each read, before it can start, has to obtain a permit. The permit has two components, a count (1) and a memory estimation. Regardless of how sophisticated this memory estimation is, the fact of life is that we *will* sometimes admit more readers than we should have, their total memory consumption surpassing the limit we tried to keep it under. We should allow the system to auto-correct the estimates dynamically, depending on the actual workload. Although the node should be resilient to mistakes like this, it is always cheaper to just admit fever reads than to deal with the consequences of admitting too many of them.

A possible solution is to introduce a *panic factor* with which we multiply our estimates. This panic factor would be adjusted based on the precision of past estimates and would decay with time (back to 1).

Bonus(?): make this panic factor tunable from the outside to allow field-engineers to intervene on a bad-behaving node.

See the entire discussion at: https://groups.google.com/forum/#!msg/scylladb-dev/SSpU-oEpNrE/ENEwopImBAAJ

Part of #4077.

**Comment by bhalevy** [comment]
According to @denesb the work we've already done in the reader concurrency semaphore area is sufficient to prevent OOM caused by readers, so we can deprioritize this (and similar, theoretical) issues.
I'll let him fill-in the details :)

**Comment by denesb** [comment]
Since 3d816b7c1 it is even harder to estimate how much resources a read is going to consume up-front as we don't know whether it will hit the cache or whether it will miss. Said work also changed how reads are admitted: they are now admitted one-by-one, the semaphore only admitting new reads if all current ones are waiting on disk or a remote shard.
Currently admission serves the purpose of not admitting new reads after we run out of resources, as opposed to trying to estimate how much resources the to-be-admitted read is going to consume and only admitting it if this is available. We have no proof that this method is causing problems, so I'm closing this.

**Comment by avikivity** [comment]
We could keep running statistics per prepared statement ID, and then seed the estimation from the statistics.

But I don't think we need to do anything now. Things look pretty stable on this front.
